File: src/Test/ArrayOpsFFI.go
package Test_ArrayOpsFFI

func arrayRange(start, end int) []int {
	step := 1
	if end < start {
		step = -1
	}
	arr := make([]int, 0)
	for i := start; ; i += step {
		arr = append(arr, i)
		if i == end {
			return arr
		}
	}
}

func arrayFilter[A any](predicate func(A) bool, arr []A) []A {
	result := make([]A, 0)
	for _, x := range arr {
		if predicate(x) {
			result = append(result, x)
		}
	}
	return result
}

func arrayFoldl[A, B any](f func(B) func(A) B, initial B, arr []A) B {
	acc := initial
	for _, x := range arr {
		acc = f(acc)(x)
	}
	return acc
}

func RunArrayOpsFFI(limit int) int {
	values := arrayRange(1, limit)
	evens := arrayFilter(func(x int) bool { return x%2 == 0 }, values)
	return arrayFoldl(func(acc int) func(int) int {
		return func(x int) int { return acc + x }
	}, 0, evens)
}

File: src/Test/ArrayOpsFFI.rs
fn range(start: i64, end: i64) -> Vec<i64> {
    if start <= end {
        (start..=end).collect()
    } else {
        (end..=start).rev().collect()
    }
}

fn filter<A: Clone>(predicate: impl Fn(&A) -> bool, array: &[A]) -> Vec<A> {
    array.iter().filter(|value| predicate(value)).cloned().collect()
}

pub fn Test_ArrayOpsFFI_runArrayOpsFFI(limit: i64) -> i64 {
    let array = range(1, limit);
    let evens = filter(|value| value % 2 == 0, &array);
    evens.iter().fold(0, |acc, value| acc + value)
}
